File: packages/sdk/javascript/src/tap/rfc9421.ts
import * as ed25519 from '@noble/ed25519';
import * as crypto from 'crypto';

export type SignatureAlgorithm = 'ed25519' | 'rsa-pss-sha256';

export interface SignatureParams {
  created: number;
  expires: number;
  keyId: string;
  alg: SignatureAlgorithm;
  nonce: string;
  tag: 'agent-browser-auth' | 'agent-payer-auth';
}

export interface SignatureComponents {
  authority: string;
  path: string;
}

export interface SignatureResult {
  signatureInput: string;
  signature: string;
}

export class RFC9421Signature {
  static createSignatureBase(
    components: SignatureComponents,
    params: SignatureParams
  ): string {
    const lines: string[] = [];

    lines.push(`"@authority": ${components.authority}`);
    lines.push(`"@path": ${components.path}`);

    const signatureParamsValue = `("@authority" "@path"); created=${params.created}; expires=${params.expires}; keyid="${params.keyId}"; alg="${params.alg}"; nonce="${params.nonce}"; tag="${params.tag}"`;
    lines.push(`"@signature-params": ${signatureParamsValue}`);

    return lines.join('\n');
  }

  static async signEd25519(
    components: SignatureComponents,
    params: SignatureParams,
    privateKey: Uint8Array
  ): Promise<SignatureResult> {
    if (privateKey.length !== 32) {
      throw new Error('Ed25519 private key must be 32 bytes');
    }

    const signatureBase = this.createSignatureBase(components, params);
    const message = new TextEncoder().encode(signatureBase);

    const signature = await ed25519.sign(message, privateKey);
    const signatureB64 = Buffer.from(signature).toString('base64');

    const signatureInput = `sig2=("@authority" "@path"); created=${params.created}; expires=${params.expires}; keyid="${params.keyId}"; alg="${params.alg}"; nonce="${params.nonce}"; tag="${params.tag}"`;

    return {
      signatureInput,
      signature: `sig2=:${signatureB64}:`,
    };
  }

  static async signRsaPss(
    components: SignatureComponents,
    params: SignatureParams,
    privateKeyPem: string
  ): Promise<SignatureResult> {
    const signatureBase = this.createSignatureBase(components, params);

    const sign = crypto.createSign('RSA-SHA256');
    sign.update(signatureBase);
    sign.end();

    const signature = sign.sign(
      {
        key: privateKeyPem,
        padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
        saltLength: crypto.constants.RSA_PSS_SALTLEN_MAX_SIGN,
      },
      'base64'
    );

    const signatureInput = `sig2=("@authority" "@path"); created=${params.created}; expires=${params.expires}; keyid="${params.keyId}"; alg="${params.alg}"; nonce="${params.nonce}"; tag="${params.tag}"`;

    return {
      signatureInput,
      signature: `sig2=:${signature}:`,
    };
  }

  static generateNonce(): string {
    const bytes = crypto.randomBytes(16);
    return bytes.toString('hex');
  }

  static generateEd25519KeyPair(): { privateKey: Uint8Array; publicKey: Uint8Array } {
    const privateKey = ed25519.utils.randomPrivateKey();
    const publicKey = ed25519.getPublicKey(privateKey);
    return { privateKey, publicKey };
  }

  static generateRsaKeyPair(): { privateKey: string; publicKey: string } {
    const { privateKey, publicKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 2048,
      publicKeyEncoding: {
        type: 'spki',
        format: 'pem',
      },
      privateKeyEncoding: {
        type: 'pkcs8',
        format: 'pem',
      },
    });

    return { privateKey, publicKey };
  }
}

File: packages/sdk/python/x402_upl/tap/rfc9421.py
import base64
import secrets
from dataclasses import dataclass
from typing import Literal, Tuple

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives import serialization


SignatureAlgorithm = Literal["ed25519", "rsa-pss-sha256"]


@dataclass
class SignatureParams:
    created: int
    expires: int
    key_id: str
    alg: SignatureAlgorithm
    nonce: str
    tag: Literal["agent-browser-auth", "agent-payer-auth"]


@dataclass
class SignatureComponents:
    authority: str
    path: str


@dataclass
class SignatureResult:
    signature_input: str
    signature: str


class RFC9421Signature:
    @staticmethod
    def create_signature_base(
        components: SignatureComponents,
        params: SignatureParams
    ) -> str:
        lines = []

        lines.append(f'"@authority": {components.authority}')
        lines.append(f'"@path": {components.path}')

        signature_params_value = (
            f'("@authority" "@path"); '
            f'created={params.created}; '
            f'expires={params.expires}; '
            f'keyid="{params.key_id}"; '
            f'alg="{params.alg}"; '
            f'nonce="{params.nonce}"; '
            f'tag="{params.tag}"'
        )
        lines.append(f'"@signature-params": {signature_params_value}')

        return '\n'.join(lines)

    @staticmethod
    async def sign_ed25519(
        components: SignatureComponents,
        params: SignatureParams,
        private_key: Ed25519PrivateKey
    ) -> SignatureResult:
        signature_base = RFC9421Signature.create_signature_base(components, params)
        message = signature_base.encode('utf-8')

        signature = private_key.sign(message)
        signature_b64 = base64.b64encode(signature).decode('utf-8')

        signature_input = (
            f'sig2=("@authority" "@path"); '
            f'created={params.created}; '
            f'expires={params.expires}; '
            f'keyid="{params.key_id}"; '
            f'alg="{params.alg}"; '
            f'nonce="{params.nonce}"; '
            f'tag="{params.tag}"'
        )

        return SignatureResult(
            signature_input=signature_input,
            signature=f'sig2=:{signature_b64}:'
        )

    @staticmethod
    def generate_nonce() -> str:
        return secrets.token_hex(16)

    @staticmethod
    def generate_ed25519_keypair() -> Tuple[Ed25519PrivateKey, Ed25519PublicKey]:
        private_key = Ed25519PrivateKey.generate()
        public_key = private_key.public_key()
        return private_key, public_key

    @staticmethod
    def private_key_to_base64(private_key: Ed25519PrivateKey) -> str:
        private_bytes = private_key.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption()
        )
        return base64.b64encode(private_bytes).decode('utf-8')

    @staticmethod
    def private_key_from_base64(b64_str: str) -> Ed25519PrivateKey:
        private_bytes = base64.b64decode(b64_str)
        return Ed25519PrivateKey.from_private_bytes(private_bytes)

    @staticmethod
    def public_key_to_base64(public_key: Ed25519PublicKey) -> str:
        public_bytes = public_key.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw
        )
        return base64.b64encode(public_bytes).decode('utf-8')
